gcc noreturn warning

Sometimes GCC thinks that a function can return when it actually can't, e.g. because it invokes a `noreturn` callback:

```
akuli@akuli-desktop:~/oomph$ cat foo.c
#include <stdio.h>
#include <stdnoreturn.h>

noreturn void foo(void){
	printf("Lol\n");
}

akuli@akuli-desktop:~/oomph$ gcc -c foo.c
foo.c: In function ‘foo’:
foo.c:6:1: warning: ‘noreturn’ function does return
 }
 ^
```

This warning makes tests fail.

--- a/self_hosted/c_output.oomph
+++ b/self_hosted/c_output.oomph
@@ -363,10 +363,12 @@ class FunctionEmitter(
                 let argvars = fdef.argvars
                 let functype = fdef.var.type as ir::FunctionType
                 let body = fdef.body
+                let function_name = fdef.var.name
             case ir::MethodDef mdef:
                 argvars = mdef.argvars
                 functype = mdef.type
                 body = mdef.body
+                function_name = ir::type_name(mdef.type.argtypes.first()) + "." + mdef.name
 
         foreach var of argvars:
             self.add_local_var(var, false)
@@ -384,7 +386,12 @@ class FunctionEmitter(
                 {self.file_pair.emit_type(type)} retval;
                 """
                 self.after_body = self.after_body + "return retval;\n"
-            case *:
+            case ir::NoReturn _:
+                let name_string = self.file_pair.emit_string(function_name)
+                self.after_body = self.after_body + """
+                panic_printf("noreturn was used, but %s() still returned", string_to_cstr({name_string}));
+                """
+            case null _:
                 pass
 
         # TODO: couldn't get list comprehension to work

--- a/tests/bad_noreturn_func_error.oomph
+++ b/tests/bad_noreturn_func_error.oomph
@@ -1,0 +1,5 @@
+func foo() -> noreturn:
+    pass
+
+export func main():
+    foo()

--- a/tests/bad_noreturn_method_error.oomph
+++ b/tests/bad_noreturn_method_error.oomph
@@ -1,0 +1,6 @@
+class Foo():
+    meth bar() -> noreturn:
+        pass
+
+export func main():
+    new Foo().bar()

--- a/tests/output/self_hosted/bad_noreturn_func_error.txt
+++ b/tests/output/self_hosted/bad_noreturn_func_error.txt
@@ -1,0 +1,2 @@
+tests/.oomph-cache/.../bad_noreturn_func_error: noreturn was used, but foo() still returned
+Program exited with status 1

--- a/tests/output/self_hosted/bad_noreturn_method_error.txt
+++ b/tests/output/self_hosted/bad_noreturn_method_error.txt
@@ -1,0 +1,2 @@
+tests/.oomph-cache/.../bad_noreturn_method_error: noreturn was used, but Foo.bar() still returned
+Program exited with status 1

--- a/tests/pyoomph_skip.txt
+++ b/tests/pyoomph_skip.txt
@@ -1,5 +1,7 @@
 # Tests for self-hosted compiler only
 tests/andor_error.oomph
+tests/bad_noreturn_func_error.oomph    # pyoomph doesn't really support noreturn
+tests/bad_noreturn_method_error.oomph  # pyoomph doesn't really support noreturn
 tests/equals_chain_error.oomph
 tests/mapping_not_hashable_error.oomph  # output depends on c compiler
 tests/mod_chain_error.oomph

--- a/tests/self_hosted_skip.txt
+++ b/tests/self_hosted_skip.txt
@@ -6,9 +6,6 @@ tests/eq_error.oomph
 tests/int_eq_str_error.oomph
 tests/optional_compare_error.oomph
 
-# https://github.com/Akuli/oomph/issues/259
-tests/noreturn_callback_bug_error.oomph
-
 # Union members are not sorted properly, it causes some problems
 tests/as_widening.oomph
 tests/callback_parens.oomph

